src/classes/ADDR_Addresses_TDTM.cls

```diff
@@ -67,231 +82,231 @@ public class ADDR_Addresses_TDTM extends TDTM_Runnable {
+	        // BEFORE INSERT
+	        if (!alreadyRunBeforeInsert && triggerAction == TDTM_Runnable.Action.BeforeInsert) {
+		        for (SObject so : listNew) {
+		            Address__c addr = (Address__c)so;
+	                // when an address is created as Default, we update its latest date fields
+	                if (addr.Default_Address__c) {
+	                    addr.Latest_Start_Date__c = System.today();
+	                    addr.Latest_End_Date__c = null;
 	               }
-	            }
-	            
-	            // BEFORE UPDATE
-	            if (triggerAction == TDTM_Runnable.Action.BeforeUpdate) {
+		        }
+		        alreadyRunBeforeInsert = true;
+	        }
+	        
+	        // BEFORE UPDATE
+	        if (!alreadyRunBeforeUpdate && triggerAction == TDTM_Runnable.Action.BeforeUpdate) {
+	            Integer i = 0;
+		        for (SObject so : listNew) {
+		            Address__c addr = (Address__c)so;
+		            Address__c addrOld = listOld == null? null : (Address__c)listOld[i];
 	                if (addr.Default_Address__c != addrOld.Default_Address__c) {
-	                   // when an address changes Default, we update its latest date fields
+	                   // when a Default address changes, we update its latest date fields
 	                    if (addr.Default_Address__c) {
 	                       addr.Latest_Start_Date__c = System.today();
 	                       addr.Latest_End_Date__c = null;
 	                    } else {
+	                       UTIL_Debug.debug('****Address updated to not default');
 	                       addr.Latest_End_Date__c = System.today();                        
 	                    }
 	                }
+	                i++;
+		        }
+		        alreadyRunBeforeUpdate = true;
+	        }
+	    
+	        // AFTER INSERT
+	        if (!alreadyRunAfterInsert && triggerAction == TDTM_Runnable.Action.AfterInsert) {
+	            for (SObject so : listNew) {
+	                Address__c addr = (Address__c)so;
+		            // a new address that is marked default or seasonal needs to propogate to the parent Account
+		            if (addr.Default_Address__c || ADDR_Seasonal_UTIL.isSeasonalAddr(addr)) {
+		                putAddrInMap(addr, mapParentIdAddr);
+		            }
+	            }
+	            
+	            if(mapParentIdAddr.size() > 0) {
+	                refreshCurrentAccAddress(mapParentIdAddr, dmlWrapper, triggerAction);
+	                processDml(dmlWrapper, mapParentIdAddr);
 	            }
-	
-	            // AFTER INSERT
-	            if (triggerAction == TDTM_Runnable.Action.AfterInsert) {
-	                // a new address that is marked default or seasonal needs to propogate to the household
-	                if (addr.Default_Address__c || isSeasonalAddr(addr)) 
-	                    mapAccIdAddr.put(addr.Parent_Account__c, addr);
+	            
+	            // enforce address records created only as children of parents
```

Comment:
should these validation checks be done before you potentially update you accounts?